utils/tinystr/src/ascii.rs

```diff
@@ -216,6 +216,67 @@ impl<const N: usize> TinyAsciiStr<N> {
         })
     }
 
+    /// Creates a `TinyAsciiStr<N>` containing the decimal representation of
+    /// the given unsigned integer.
+    ///
+    /// If the number of decimal digits exceeds `N`, the highest-magnitude
+    /// digits are truncated, and the lowest-magnitude digits are returned
+    /// as the error.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tinystr::TinyAsciiStr;
+    /// use tinystr::tinystr;
+    ///
+    /// let s0_4 = TinyAsciiStr::<4>::new_decimal_usize(0).unwrap();
+    /// let s456_4 = TinyAsciiStr::<4>::new_decimal_usize(456).unwrap();
+    /// let s456_3 = TinyAsciiStr::<3>::new_decimal_usize(456).unwrap();
+    /// let s456_2 = TinyAsciiStr::<2>::new_decimal_usize(456).unwrap_err();
+    ///
+    /// assert_eq!(s0_4, tinystr!(4, "0"));
+    /// assert_eq!(s456_4, tinystr!(4, "456"));
+    /// assert_eq!(s456_3, tinystr!(3, "456"));
+    /// assert_eq!(s456_2, tinystr!(2, "56"));
+    /// ```
+    ///
+    /// Example with saturating the value:
+    ///
+    /// ```
+    /// use tinystr::TinyAsciiStr;
+    /// use tinystr::tinystr;
+    ///
+    /// let str_truncated = TinyAsciiStr::<2>::new_decimal_usize(456).unwrap_or_else(|s| s);
+    /// let str_saturated = TinyAsciiStr::<2>::new_decimal_usize(456).unwrap_or(tinystr!(2, "99"));
+    ///
+    /// assert_eq!(str_truncated, tinystr!(2, "56"));
+    /// assert_eq!(str_saturated, tinystr!(2, "99"));
+    /// ```
+    pub fn new_decimal_usize(number: usize) -> Result<Self, Self> {
+        let mut bytes = [0; N];
+        let mut x = number;
+        let mut i = 0usize;
+        #[expect(clippy::indexing_slicing)] // in-range: i < N
+        while i < N && (x != 0 || i == 0) {
+            bytes[N - i - 1] = ((x % 10) as u8) + b'0';
+            x /= 10;
+            i += 1;
+        }
+        if i < N {
+            bytes.copy_within((N - i)..N, 0);
+            bytes[i..N].fill(0);
+        }
+        let s = Self {
+            // SAFETY: `out` only contains ASCII bytes and has same size as `self.bytes`
+            bytes: unsafe { AsciiByte::to_ascii_byte_array(&bytes) },
```

Comment:
issue: this is a complex function, and the invariant is hard to track

Instead, we can write this function safely, and have it call `AsciiByte::from_digit()`. `from_digit()` has a debug-assertion invariant that it is given a digit, and it does a bitmask-and-add to produce an ASCII byte.


You can also write it fully safely using a `match`, and it _should_ compile to a jump table, but it's understandable if you're not sure of that. Play around with it in godbolt, perhaps.